xdrv_21_wemo.ino

```diff
@@ -0,0 +1,486 @@
+
+    LogUpnpWithClient(PSTR(D_WEMO_EVENT_SERVICE));
+
+#ifdef USE_UNISHOX_COMPRESSION
+    InternalWSSend(200, CT_PLAIN, Decompress(WEMO_EVENTSERVICE_XML, WEMO_EVENTSERVICE_XML_SIZE));
+#else
+    InternalWSSend(200, CT_PLAIN, FPSTR(WEMO_EVENTSERVICE_XML));
+#endif
+  }
+
+  void HandleUpnpMetaService(void)
+  {
+    AddLog_P2(LOG_LEVEL_DEBUG, PSTR("\nHandleUpnpMetaService"));
+
+    LogUpnpWithClient(PSTR(D_WEMO_META_SERVICE));
+
+#ifdef USE_UNISHOX_COMPRESSION
+    InternalWSSend(200, CT_PLAIN, Decompress(WEMO_METASERVICE_XML, WEMO_METASERVICE_XML_SIZE));
+#else
+    InternalWSSend(200, CT_PLAIN, FPSTR(WEMO_METASERVICE_XML));
+#endif
+  }
+
+  void HandleUpnpSetupWemo(void)
+  {
+    LogUpnpWithClient(PSTR(D_WEMO_SETUP));
+
+#ifdef USE_UNISHOX_COMPRESSION
+    String setup_xml = Decompress(WEMO_SETUP_XML, WEMO_SETUP_XML_SIZE);
+#else
+    String setup_xml = FPSTR(WEMO_SETUP_XML);
+#endif
+    setup_xml.replace("{x1", SettingsText(SET_FRIENDLYNAME1 + (_deviceId - 1)));
+    setup_xml.replace("{x2", WemoUuid());
+    setup_xml.replace("{x3", WemoSerialnumber());
+    InternalWSSend(200, CT_XML, setup_xml);
+
+    AddLog_P2(LOG_LEVEL_DEBUG, PSTR("\nSENDING DEVICE #%d: %s"), _deviceId, setup_xml.c_str());
+  }
+
+public:
+  void RegisterHandlers(void)
+  {
+    AddLog_P2(LOG_LEVEL_DEBUG, PSTR("\nREGISTER DEVICE #%d"), _deviceId);
+
+    _webServer->on(F("/upnp/control/basicevent1"), [&]() { HandleUpnpEvent(); });
+    _webServer->on(F("/eventservice.xml"), [&]() { HandleUpnpService(); });
+    _webServer->on(F("/metainfoservice.xml"), [&]() { HandleUpnpMetaService(); });
+    _webServer->on(F("/setup.xml"), [&]() { HandleUpnpSetupWemo(); });
+  }
+};
+
+WemoSwitch *wemoDevice[MAX_FRIENDLYNAMES] = {};
+int numOfWemoSwitch = 0;
+
+void WemoRespondToMSearch(int echo_type)
+{
+  TickerMSearch.detach();
+
+  for (int i = 0; i < numOfWemoSwitch; i++)
+  {
+    wemoDevice[i]->WemoRespondToMSearch(echo_type);
+  }
+
+  udp_response_mutex = false;
+}
+
+/*********************************************************************************************\
+ * Interface
+\*********************************************************************************************/
+
+bool Xdrv21(uint8_t function)
+{
+  bool result = false;
+
+  if (devices_present && (EMUL_WEMO == Settings.flag2.emulation))
+  {
+    switch (function)
+    {
+    case FUNC_LOOP:
+      for (int i = 1; i < numOfWemoSwitch; i++) // Handle devices web server
+      {
+        wemoDevice[i]->HandleServerLoop();
+      }
+      break;
+    case FUNC_WEB_ADD_HANDLER:
+
+      AddLog_P2(LOG_LEVEL_DEBUG, PSTR("\nADDING HANDLERS FOR %d DEVICES"), devices_present);
+
+      // For the first device use the current webserver, for the others.. create a new one listening on a different PortUdp
+      wemoDevice[numOfWemoSwitch] = new WemoSwitch(1, Webserver);
+      wemoDevice[numOfWemoSwitch]->RegisterHandlers();
+      numOfWemoSwitch++;
+
+      for (int i = 1; i < devices_present; i++)
```

Comment:
Don't you have numOfWemoSwitch == devices_present in the end?

Be careful that devices_present will also include PWMs. If you want only the number of relays, `Light.device`. I'm no sure whether this survives if you undef `USE_LIGHT` though.

Reply:
If all's went as it should, device_present should be equal at numOfWemoSwitch.
I din't know about Light.device. This is my first time that I work on Arduino.

Reply:
No problem, your code is excellent. devices_present will list the number of devices that can receive on/off commands. It could make sense to keep it, you would also be able to turn lights on/off but not other parameters.

Reply:
For Lights management, I would use HUE emulation. I would change the config description from:
- WEMO (Single device)
- HUE Emulation (Multi device)
into something like:
- HUE Emulation (for lights)
- WEMO (for switches)
I don't know if there are other kind of devices, but just to give the idea). Lights should be used with HUE emu.